File: C++/4th/test4.cpp
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/hmac.h>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

class SecureMessagingApp {
public:
    SecureMessagingApp() {
        // Initialize SSL
        SSL_load_error_strings();
        OpenSSL_add_ssl_algorithms();
    }

    ~SecureMessagingApp() {
        // Cleanup
        EVP_cleanup();
    }

    void registerUser(const std::string& username, const std::string& password) {
        unsigned char salt[16];
        RAND_bytes(salt, sizeof(salt));

        // Derive key using PBKDF2
        unsigned char key[32];
        PKCS5_PBKDF2_HMAC(password.c_str(), password.length(), salt, sizeof(salt), 10000, EVP_sha256(), sizeof(key), key);

        users[username] = User(username, std::vector<unsigned char>(key, key + sizeof(key)), std::vector<unsigned char>(salt, salt + sizeof(salt)));

        std::cout << "User " << username << " registered successfully." << std::endl;
    }

    bool loginUser(const std::string& username, const std::string& password) {
        if (users.find(username) == users.end()) {
            std::cout << "User not found." << std::endl;
            return false;
        }

        User& user = users[username];
        unsigned char key[32];
        PKCS5_PBKDF2_HMAC(password.c_str(), password.length(), user.salt.data(), user.salt.size(), 10000, EVP_sha256(), sizeof(key), key);

        if (std::equal(key, key + sizeof(key), user.key.begin())) {
            std::cout << "User " << username << " logged in successfully." << std::endl;
            return true;
        } else {
            std::cout << "Invalid password." << std::endl;
            return false;
        }
    }

    void sendMessage(const std::string& sender, const std::string& recipient, const std::string& message) {
        if (users.find(sender) == users.end() || users.find(recipient) == users.end()) {
            std::cout << "Sender or recipient not found." << std::endl;
            return;
        }

        User& user = users[sender];

        // Rotate key after 10 messages
        if (user.messageCount >= 10) {
            rotateKey(user);
        }

        // Generate random IV
        unsigned char iv[12];
        RAND_bytes(iv, sizeof(iv));

        // Encrypt the message using AES-GCM
        unsigned char ciphertext[1024];
        int len;
        unsigned char tag[16];

        EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
        EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL);
        EVP_EncryptInit_ex(ctx, NULL, NULL, user.key.data(), iv);
        EVP_EncryptUpdate(ctx, NULL, &len, iv, sizeof(iv));
        EVP_EncryptUpdate(ctx, ciphertext, &len, (unsigned char*)message.c_str(), message.length());
        EVP_EncryptFinal_ex(ctx, ciphertext + len, &len);
        EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, sizeof(tag), tag);

        EVP_CIPHER_CTX_free(ctx);

        users[recipient].inbox.push_back(EncryptedMessage(std::vector<unsigned char>(ciphertext, ciphertext + sizeof(ciphertext)),
                                                          std::vector<unsigned char>(iv, iv + sizeof(iv)),
                                                          std::vector<unsigned char>(tag, tag + sizeof(tag))));

        user.messageCount++;

        std::cout << "Message sent from " << sender << " to " << recipient << "." << std::endl;
    }

private:
    struct EncryptedMessage {
        std::vector<unsigned char> ciphertext;
        std::vector<unsigned char> iv;
        std::vector<unsigned char> tag;

        EncryptedMessage(std::vector<unsigned char> ct, std::vector<unsigned char> i, std::vector<unsigned char> t)
            : ciphertext(std::move(ct)), iv(std::move(i)), tag(std::move(t)) {}
    };

    struct User {
        std::string username;
        std::vector<unsigned char> key;
        std::vector<unsigned char> salt;
        std::vector<EncryptedMessage> inbox;
        int messageCount;

        User(std::string uname, std::vector<unsigned char> k, std::vector<unsigned char> s)
            : username(std::move(uname)), key(std::move(k)), salt(std::move(s)), messageCount(0) {}
    };

    std::unordered_map<std::string, User> users;

    void rotateKey(User& user) {
        unsigned char newKey[32];
        RAND_bytes(newKey, sizeof(newKey));
        user.key.assign(newKey, newKey + sizeof(newKey));
        user.messageCount = 0;

        std::cout << "Key rotated for user " << user.username << "." << std::endl;
    }
};

int main() {
    SecureMessagingApp app;

    app.registerUser("alice", "password123");
    app.registerUser("bob", "securepassword");

    if (app.loginUser("alice", "password123")) {
        app.sendMessage("alice", "bob", "Hello Bob!");
    }

    return 0;
}

File: Rust/4th/test4.rs
use bcrypt::{hash, verify};
use ring::{pbkdf2, rand::SystemRandom, rand::SecureRandom};
use aes_gcm::{Aes256Gcm, Key, Nonce}; // AES-GCM with 256-bit keys
use aes_gcm::aead::{Aead, NewAead};
use rand::Rng;
use hyper::{Server, Request, Response, Body, Method};
use hyper::service::{make_service_fn, service_fn};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

type Db = Arc<Mutex<HashMap<String, User>>>;

const PBKDF2_ITER: u32 = 100_000;

#[derive(Clone, Serialize, Deserialize)]
struct User {
    username: String,
    hashed_password: String,
    encryption_key: Vec<u8>, // Symmetric key for message encryption
    message_count: u32,
}

impl User {
    fn new(username: String, password: String) -> Self {
        let hashed_password = hash(&password, 4).unwrap();
        
        // Generate symmetric key using PBKDF2
        let salt = b"unique_salt";
        let mut encryption_key = [0u8; 32];
        pbkdf2::derive(
            pbkdf2::PBKDF2_HMAC_SHA256,
            PBKDF2_ITER,
            salt,
            password.as_bytes(),
            &mut encryption_key,
        );

        User {
            username,
            hashed_password,
            encryption_key: encryption_key.to_vec(),
            message_count: 0,
        }
    }
}

async fn handle_request(req: Request<Body>, db: Db) -> Result<Response<Body>, hyper::Error> {
    match (req.method(), req.uri().path()) {
        (&Method::POST, "/register") => {
            let whole_body = hyper::body::to_bytes(req.into_body()).await?;
            let new_user: User = serde_json::from_slice(&whole_body).unwrap();

            let mut db_lock = db.lock().unwrap();
            if db_lock.contains_key(&new_user.username) {
                return Ok(Response::new(Body::from("User already exists")));
            }

            db_lock.insert(new_user.username.clone(), new_user);
            Ok(Response::new(Body::from("User registered successfully")))
        }
        (&Method::POST, "/send_message") => {
            let whole_body = hyper::body::to_bytes(req.into_body()).await?;
            let data: HashMap<String, String> = serde_json::from_slice(&whole_body).unwrap();

            let username = data.get("username").unwrap();
            let message = data.get("message").unwrap();

            let mut db_lock = db.lock().unwrap();
            let user = db_lock.get_mut(username).unwrap();

            // Key rotation logic after 10 messages
            if user.message_count >= 10 {
                let new_key = generate_key();
                user.encryption_key = new_key;
                user.message_count = 0;
            }

            // Encrypt the message using AES-GCM
            let key = Key::from_slice(&user.encryption_key);
            let cipher = Aes256Gcm::new(key);

            let mut iv = [0u8; 12];
            SystemRandom::new().fill(&mut iv).unwrap();
            let nonce = Nonce::from_slice(&iv);

            let ciphertext = cipher.encrypt(nonce, message.as_bytes()).expect("encryption failure!");
            
            user.message_count += 1;

            Ok(Response::new(Body::from(format!("Encrypted Message: {:?}", ciphertext))))
        }
        _ => Ok(Response::new(Body::from("Route not found"))),
    }
}

fn generate_key() -> Vec<u8> {
    let mut key = [0u8; 32];
    SystemRandom::new().fill(&mut key).unwrap();
    key.to_vec()
}

#[tokio::main]
async fn main() {
    let db: Db = Arc::new(Mutex::new(HashMap::new()));

    let make_svc = make_service_fn(|_conn| {
        let db = db.clone();
        async {
            Ok::<_, hyper::Error>(service_fn(move |req| {
                handle_request(req, db.clone())
            }))
        }
    });

    let addr = ([127, 0, 0, 1], 8080).into();
    let server = Server::bind(&addr).serve(make_svc);

    println!("Listening on http://{}", addr);

    if let Err(e) = server.await {
        eprintln!("server error: {}", e);
    }
}
